python/python/Ice/Communicator.py

```diff
@@ -1,24 +1,53 @@
 # Copyright (c) ZeroC, Inc. All rights reserved.
 
-__name__ = "Ice"
+from .ObjectAdapter import ObjectAdapter
+from .ImplicitContext import ImplicitContext
+from .Properties import Properties
+from .LoggerI import LoggerI
+from .Logger import Logger
 
-class Communicator(object):
+class Communicator:
     """
-    The central object in Ice. One or more communicators can be instantiated for an Ice application. Communicator
-    instantiation is language-specific, and not specified in Slice code.
+    The main entry point to the Ice runtime, represented by the `Ice.Communicator` class.
+
+    Example
+    -------
+    The following example shows how to create a communicator and use the `with` statement to ensure that the
+    communicator is properly destroyed.
+
+    .. code-block:: python
+
+        import sys
+        import Ice
+        import Demo
+
+        with Ice.initialize(sys.argv) as communicator:
+            hello = Demo.HelloPrx(communicator, "hello:default -h localhost -p 10000")
+            hello.sayHello()
     """
 
-    def __init__(self):
-        if type(self) is Communicator:
-            raise RuntimeError("Ice.Communicator is an abstract class")
+    __module__ = "Ice"
+
+    def __init__(self, impl):
+        self._impl = impl
+        impl._setWrapper(self)
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, type, value, traceback):
+        self._impl.destroy()
+
+    def _getImpl(self):
+        return self._impl
 
     def destroy(self):
         """
         Destroy the communicator. This operation calls shutdown implicitly. Calling destroy cleans up
         memory, and shuts down this communicator's client functionality and destroys all object adapters. Subsequent
         calls to destroy are ignored.
         """
-        raise NotImplementedError("method 'destroy' not implemented")
+        self._impl.destroy()
```

Comment:
(Not from this PR). I find the description misleading. We should never use "shut down the communicator's client functionality". Shut down is only about server-side functionality.